binding.c

```diff
@@ -11,24 +99,2109 @@ napi_value sn_randombytes_random (napi_env env, napi_callback_info info) {
+}
+
+napi_value sn_crypto_secretstream_xchacha20poly1305_pull (napi_env env, napi_callback_info info) {
+  SN_ARGV(4, crypto_secretstream_xchacha20poly1305_pull)
+
+  SN_ARGV_BUFFER_CAST(crypto_secretstream_xchacha20poly1305_state *, state, 0)
+  SN_ARGV_TYPEDARRAY(message, 1)
+  SN_ARGV_TYPEDARRAY(ciphertext, 2)
+  SN_ARGV_OPTS_TYPEDARRAY(ad, 3)
+
+  SN_THROWS(state_size != sizeof(crypto_secretstream_xchacha20poly1305_state), "state must be 52 bytes")
+  SN_THROWS(message_size != ciphertext_size - crypto_secretstream_xchacha20poly1305_ABYTES, "message must be less than 17 bytes")
+  SN_THROWS(ciphertext_size != message_size + crypto_secretstream_xchacha20poly1305_ABYTES, "key must 24 bytes")
+
+  SN_RETURN(crypto_secretstream_xchacha20poly1305_pull(state, message_data, NULL, NULL, ciphertext_data, ciphertext_size, ad_data, ad_size), "pull failed")
+}
+
+napi_value sn_crypto_secretstream_xchacha20poly1305_rekey (napi_env env, napi_callback_info info) {
+  SN_ARGV(1, crypto_secretstream_xchacha20poly1305_rekey)
+
+  SN_ARGV_BUFFER_CAST(crypto_secretstream_xchacha20poly1305_state *, state, 0)
+
+  SN_THROWS(state_size != sizeof(crypto_secretstream_xchacha20poly1305_state), "state must be 52 bytes")
+
+  crypto_secretstream_xchacha20poly1305_rekey(state);
+
+  return NULL;
+}
+
+// TODO: SN_ARGV_UINT8 not working
+
+// async
+typedef struct async_pwhash_request {
+  napi_ref output_ref;
+  unsigned char * output_data;
+  size_t output_size;
+  napi_ref pwd_ref;
+  const char * pwd_data;
+  size_t pwd_size;
+  napi_ref salt_ref;
+  unsigned char * salt;
+  uint32_t opslimit;
+  uint32_t memlimit;
+  uint32_t algorithm;
+  napi_ref cb;
+  int n;
+  napi_async_work task;
+} async_pwhash_request;
+
+static void async_pwhash_execute(napi_env env, void* req_v) {
+  struct async_pwhash_request * req = (async_pwhash_request *)req_v;
+  req->n = crypto_pwhash(req->output_data, req->output_size, req-> pwd_data, req->pwd_size, req->salt, req->opslimit, req->memlimit, req->algorithm);
+}
+
+static void async_pwhash_complete(napi_env env, napi_status status, void* data) {
+  async_pwhash_request * req = (async_pwhash_request *)data;
+  SN_STATUS_THROWS_VOID(status, "");
+
+  napi_value global;
+  SN_STATUS_THROWS_VOID(napi_get_global(env, &global), "");
+
+  napi_value argv[1];
+  napi_get_null(env, &argv[0]);
+
+  napi_value callback;
+  SN_STATUS_THROWS_VOID(napi_get_reference_value(env, req->cb, &callback), "");
+
+  napi_value return_val;
+  SN_STATUS_THROWS_VOID(napi_call_function(env, global, callback, 1, argv, &return_val), "");
+  SN_STATUS_THROWS_VOID(napi_delete_reference(env, req->cb), "");
+  SN_STATUS_THROWS_VOID(napi_delete_reference(env, req->output_ref), "");
+  SN_STATUS_THROWS_VOID(napi_delete_reference(env, req->pwd_ref), "");
+  SN_STATUS_THROWS_VOID(napi_delete_reference(env, req->salt_ref), "");
+  SN_STATUS_THROWS_VOID(napi_delete_async_work(env, req->task), "");
```

Comment:
none of these can throw afaik as it's async. should just be asserts

Reply:
fixed by f228156